### ISSUE TITLE
gperf-3.1 lookup functions take size_t length parameter instead of unsigned int

### ISSUE THREAD

**Issue by floppym** [systemd/systemd#5039]
### Submission type

  - [x] Bug report
  - [ ] Request for enhancement (RFE)

*NOTE: Do not submit anything other than bug reports or RFEs via the issue tracker!*

### systemd version the issue has been seen with

232

### Used distribution

Gentoo

### Summary

Downstream bug: https://bugs.gentoo.org/show_bug.cgi?id=605022

After upgrading to gperf-3.1, systemd fails to build due to type mismatches for arguments of `lookup_af` and `lookup_errno`.

```
In file included from /var/tmp/portage/sys-apps/systemd-232/work/systemd-232/src/basic/af-list.c:28:0:
./src/basic/af-from-name.h:140:1: error: conflicting types for ‘lookup_af’
 lookup_af (register const char *str, register size_t len)
 ^~~~~~~~~
/var/tmp/portage/sys-apps/systemd-232/work/systemd-232/src/basic/af-list.c:26:30: note: previous declaration of ‘lookup_af’ was here
 static const struct af_name* lookup_af(register const char *str, register unsigned int len);
                              ^~~~~~~~~
make[2]: *** [Makefile:15935: src/basic/libbasic_la-af-list.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from /var/tmp/portage/sys-apps/systemd-232/work/systemd-232/src/basic/errno-list.c:28:0:
./src/basic/errno-from-name.h:143:1: error: conflicting types for ‘lookup_errno’
 lookup_errno (register const char *str, register size_t len)
 ^~~~~~~~~~~~
/var/tmp/portage/sys-apps/systemd-232/work/systemd-232/src/basic/errno-list.c:25:33: note: previous declaration of ‘lookup_errno’ was here
 static const struct errno_name* lookup_errno(register const char *str,
                                 ^~~~~~~~~~~~
make[2]: *** [Makefile:15928: src/basic/libbasic_la-errno-list.lo] Error 1
mv -f src/basic/.deps/libbasic_la-cgroup-util.Tpo src/basic/.deps/libbasic_la-cgroup-util.Plo
make[1]: *** [Makefile:21960: all-recursive] Error 1
make: *** [Makefile:9900: all] Error 2
```

The gperf NEWS file says this:

```
New in 3.1:
* The generated C code is now in ANSI-C by default. If you want to support
  pre-ANSI-C compilers, you need to provide the option --language=C on the
  command line or %language=C in the source file.
* The 'len' parameter of the hash function and of the lookup function is now
  of type 'size_t' instead of 'unsigned int'. This makes it safe to call these
  functions with strings of length > 4 GB, on 64-bit machines.
* Added option --constants-prefix.
* Added declaration %define constants-prefix.
```

**Comment by floppym** [comment]
This could be easily solved by replacing unsigned int with size_t, but that would break backward compatibility with older gperf.

Maybe there is some way to detect the gperf version at configure time?

**Comment by jsynacek** [comment]
The fix breaks compilation with gperf-3.0.x.
